From: Julian Brown <julian@codesourcery.com>
Date: Thu, 22 Oct 2009 19:22:35 +0000 (+0000)
Subject: Fix locating GOT for Thumb-2 PIE binaries.
X-Git-Tag: glibc-2.11~21
X-Git-Url: http://sourceware.org/git/?p=glibc-ports.git;a=commitdiff_plain;h=b5c2620bee9fb698c642fa609ee03d6a87ac1552

Fix locating GOT for Thumb-2 PIE binaries.

	* sysdeps/arm/elf/start.S (_start): Avoid dependency on PC pipeline
	offset.
---

diff --git a/ports/sysdeps/arm/elf/start.S b/ports/sysdeps/arm/elf/start.S
index 0cf4339..d21102e 100644
--- a/ports/sysdeps/arm/elf/start.S
+++ b/ports/sysdeps/arm/elf/start.S
@@ -87,8 +87,8 @@ _start:
 
 #ifdef SHARED
 	ldr sl, .L_GOT
-.L_GOT_OFF:
-	add sl, pc, sl
+	adr a4, .L_GOT
+	add sl, sl, a4
 
 	ldr ip, .L_GOT+4	/* __libc_csu_fini */
 	ldr ip, [sl, ip]
@@ -125,7 +125,7 @@ _start:
 
 #ifdef SHARED
 .L_GOT:
-	.word _GLOBAL_OFFSET_TABLE_-(.L_GOT_OFF+8)
+	.word _GLOBAL_OFFSET_TABLE_ - .L_GOT
 	.word __libc_csu_fini(GOT)
 	.word __libc_csu_init(GOT)
 	.word main(GOT)
